apps/protocol-frontend/src/contexts/ContributionContext.tsx

```diff
@@ -0,0 +1,590 @@
+
+const ITEMS_PER_PAGE = 100;
+
+export const ContributionContext = createContext<ContributionContextType>(
+  {} as ContributionContextType,
+);
+
+type UserContributionsDateRangeCountType = {
+  count: number;
+  date: string;
+  guildIds?: number[] | null | undefined;
+  name?: string | null | undefined;
+};
+
+interface ContributionContextProps {
+  children: React.ReactNode;
+}
+
+export const ContributionsContextProvider: React.FC<
+  ContributionContextProps
+> = ({ children }: ContributionContextProps) => {
+  const { userData } = useContext(UserContext);
+  const toast = useToast();
+  const { data: signer } = useSigner();
+  const { chain } = useNetwork();
+
+  const govrn = new GovrnProtocol(PROTOCOL_URL, { credentials: 'include' });
+  const { setModals } = useOverlay();
+
+  const [contribution, setContribution] = useState<UIContribution>(
+    {} as UIContribution,
+  );
+
+  const [userContributionPage, setUserContributionPage] = useState(0);
+  const [isUserContributionsHaveMore, setUserContributionHasMore] =
+    useState(true);
+  const [isUserContributionsLoading, setUserContributionsLoading] =
+    useState(true);
+  const [userContributions, setUserContributions] = useState<UIContribution[]>(
+    [],
+  );
+
+  const [daoContributionPage, setDaoContributionPage] = useState(0);
+  const [isDaoContributionLoading, setDaoContributionLoading] = useState(true);
+  const [isDaoContributionsHaveMore, setDaoContributionHasMore] =
+    useState(true);
+  const [daoContributions, setDaoContributions] = useState<UIContribution[]>(
+    [],
+  );
+
+  const [userAttestations, setUserAttestations] =
+    useState<UIAttestations | null>(null);
+
+  const [userContributionsDateRangeCount, setUserContributionsDateRangeCount] =
+    useState<UserContributionsDateRangeCountType[]>([]);
+
+  const userId = userData?.id;
+
+  useEffect(() => {
+    getUserContributions(userContributionPage);
+  }, [userContributionPage, userId]);
+
+  useEffect(() => {
+    getDaoContributions(daoContributionPage);
+  }, [daoContributionPage, userData?.id]);
+
+  const getContribution = async (id: number) => {
+    try {
+      const contributionResponse = await govrn.contribution.get(id);
+      if (contributionResponse) {
+        const formattedResponse = {
+          ...contributionResponse,
+          date_of_engagement: formatDate(
+            contributionResponse.date_of_engagement,
+          ),
+          date_of_submission: formatDate(
+            contributionResponse.date_of_submission,
+          ),
+        };
+        setContribution(formattedResponse);
+        return formattedResponse;
+      }
+      return null;
+    } catch (error) {
+      console.error(error);
+      return null;
+    }
+  };
+
+  const loadNextUserContributionsPage = () => {
+    setUserContributionPage(userContributionPage + 1);
+  };
+
+  const getUserContributions = async (page = 0) => {
+    setUserContributionsLoading(page === 0);
```

Comment:
Why is this page === 0 instead of true. Why wouldn't it be loading if we are fetch page 1 or 2 etc...

Reply:
Good question!
This is for the spinner outside the table. So, if it's set to `true` all time, it will hide the whole table instead. I think this makes no good UX.